server/src/main/java/org/opensearch/index/ShardIndexingPressureTracker.java

```diff
@@ -0,0 +1,223 @@
+        private final AtomicLong lastSuccessfulRequestLimitsBreachedRejections = new AtomicLong();
+        private final AtomicLong throughputDegradationLimitsBreachedRejections = new AtomicLong();
+
+        public AtomicLong getTotalRejections() {
+            return totalRejections;
+        }
+
+        public AtomicLong getNodeLimitsBreachedRejections() {
+            return nodeLimitsBreachedRejections;
+        }
+
+        public AtomicLong getLastSuccessfulRequestLimitsBreachedRejections() {
+            return lastSuccessfulRequestLimitsBreachedRejections;
+        }
+
+        public AtomicLong getThroughputDegradationLimitsBreachedRejections() {
+            return throughputDegradationLimitsBreachedRejections;
+        }
+    }
+
+    /**
+     * Performance Tracker is used to track the request performance statistics for every operation, per shard. This includes:
+     * a. latencyInMillis - Total indexing time take by requests that were processed successfully for a shard.
+     * b. lastSuccessfulRequestTimestamp - Timestamp of last successful request for a shard.
+     * c. TotalOutstandingRequests - Total requests outstanding for a shard at any given point.
+     * d. ThroughputMovingAverage - Total moving average throughput value for last N requests.
+     * e. ThroughputMovingQueue - Queue that holds the last N requests throughput such that there exists a sliding window
+     *          which keeps moving everytime a new request comes. At any given point it tracks last N requests only.
+     *          EWMA cannot be used here as it evaluate the historical average, while here it just needs the average of last N requests.
+     */
+    public static class PerformanceTracker {
+        private final AtomicLong latencyInMillis = new AtomicLong();
+        private final AtomicLong lastSuccessfulRequestTimestamp = new AtomicLong();
+        private final AtomicLong totalOutstandingRequests = new AtomicLong();
+        /**
+         * Shard Window Throughput Tracker.
+         * We will be using atomic long to track double values as mentioned here -
+         * https://docs.oracle.com/javase/6/docs/api/java/util/concurrent/atomic/package-summary.html
+         */
+        private final AtomicLong throughputMovingAverage = new AtomicLong();
+        private final ConcurrentLinkedQueue<Double> throughputMovingQueue = new ConcurrentLinkedQueue();
+
+        public AtomicLong getLatencyInMillis() {
+            return latencyInMillis;
+        }
+
+        public AtomicLong getLastSuccessfulRequestTimestamp() {
+            return lastSuccessfulRequestTimestamp;
+        }
+
+        public AtomicLong getTotalOutstandingRequests() {
+            return totalOutstandingRequests;
+        }
+
+        public AtomicLong getThroughputMovingAverage() {
+            return throughputMovingAverage;
+        }
+
+        public ConcurrentLinkedQueue<Double> getThroughputMovingQueue() {
+            return throughputMovingQueue;
+        }
```

Comment:
Should we create a class `MovingAverage` in commons package for re-usability

Reply:
Yes, we can do a follow-up on this once the logic is ready, to see if the it makes sense to move it a common utility.